### PR TITLE
Makefile fixes

### PR DESCRIPTION
**Pull request by sshedi**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,52 +12,91 @@
...
-REVISION=$(shell git rev-parse HEAD)$(shell if ! git diff --no-ext-diff --quiet --exit-code; then echo .m; fi)
+REVISION ?= $(shell $(GIT) rev-parse HEAD)$(shell if ! $(GIT) diff --no-ext-diff --quiet --exit-code; then $(ECHO) .m; fi)
 
 PKG_MAIN := github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc
 PKG_VERSION := github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc/version
 
-GO ?= go
 export GO_BUILD=GO111MODULE=on CGO_ENABLED=0 $(GO) build -ldflags "-s -w -X $(PKG_VERSION).Version=$(VERSION) -X $(PKG_VERSION).Revision=$(REVISION)"
 
-bin/containerd-fuse-overlayfs-grpc:
```
Comment (AkihiroSuda) [inline Makefile:27]: Why is this changed?

File: Makefile [hunk 2]
```diff
@@ -12,52 +12,91 @@
...
 #   limitations under the License.
 
+# Deliverables path
 DESTDIR ?= /usr/local
-
-VERSION=$(shell git describe --match 'v[0-9]*' --dirty='.m' --always --tags)
+BINDIR ?= $(DESTDIR)/bin
+
+# Tools path
+RM ?= /bin/rm
```
Comment (AkihiroSuda) [inline Makefile:20]: Isn't this implied by default?
https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html

File: Makefile [hunk 3]
```diff
@@ -12,52 +12,91 @@
...
+ECHO ?= /bin/echo
+DOCKER ?= /bin/docker
+GO ?= /bin/go
+MKDIR ?= /bin/mkdir
+TAR ?= /bin/tar
+INSTALL ?= /bin/install
+GIT ?= /bin/git
```
Comment (AkihiroSuda) [inline Makefile:27]: docker, go, git are very unlikely to be installed under `/bin`.

Why not just use the default binaries in the PATH?

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Sorry, I don't get what is improved in this PR

File: Makefile (hunk 1 above)
Comment (sshedi) [reply Makefile:27]: It's better give final binary name as target than using `make bin/containerd-fuse-overlayfs-grpc` to build target binary.

File: Makefile (hunk 2 above)
Comment (sshedi) [reply Makefile:20]: Yeah, will fix it.

File: Makefile (hunk 3 above)
Comment (sshedi) [reply Makefile:27]: How about using `GIT ?= $(which git)`

**Comment by sshedi** [comment]
> Sorry, I don't get what is improved in this PR

Multiple things.

1. Makefile is more flexible now
2. Easier to build rpms from tagged source tarball
3. Added a help target
4. Some minor tweaks for ease of maintenance.

File: Makefile (hunk 1 above)
Comment (AkihiroSuda) [reply Makefile:27]: Why?

File: Makefile (hunk 3 above)
Comment (AkihiroSuda) [reply Makefile:27]: Isn't `GIT ?= git` enough?

File: Makefile (hunk 1 above)
Comment (sshedi) [reply Makefile:27]: Ease of use, speaking generally `make <target-name>` is a standard practice. We don't use `make <path>/<target-name>`. Please let me know if you think otherwise here.

File: Makefile (hunk 3 above)
Comment (sshedi) [reply Makefile:27]: Yeah that should suffice, I though of using absolute paths. A day might come where I have multiple versions of git or docker installed in my machine and I want to use given version of docker or git for testing. But `GIT ?= git` should suffice.

I though something like this:
```
# Tools path
ECHO ?= $(shell which echo)
DOCKER ?= $(shell which docker)
GO ?= $(shell which go)
MKDIR ?= $(shell which mkdir)
TAR ?= $(shell which tar)
INSTALL ?= $(shell which install)
GIT ?= $(shell which git)
```

File: Makefile (hunk 1 above)
Comment (sshedi) [reply Makefile:27]: Else we can keep this as it is and use `make binaries` or `make all` which will implicitly call this target.
Even `make containerd-fuse-overlayfs-grpc` is a lot of characters to type. :)

Comment (AkihiroSuda) [reply Makefile:27]: The standard practice is to use file names as target names, and users are not expected to type the file names (target names) manually, as they are usually depended by the default target (such as `binaries`, `all`)

Comment (sshedi) [reply Makefile:27]: Yeah, I guess we are good with this then ???

Comment (AkihiroSuda) [reply Makefile:27]: The file name is `bin/containerd-fuse-overlayfs-grpc`, so the target name should be kept `bin/containerd-fuse-overlayfs-grpc`

File: Makefile (hunk 3 above)
Comment (AkihiroSuda) [reply Makefile:27]: `which` is often missing on recent distros

Comment (sshedi) [reply Makefile:27]: Yeah, I am also thinking same. Let's go with your approach. Thanks.

File: Makefile (hunk 1 above)
Comment (sshedi) [reply Makefile:27]: Okay

**Comment by sshedi** [comment]
Hi @AkihiroSuda I think I have addressed all your comments, please review again.

**Comment by sshedi** [comment]
Hi @AkihiroSuda, gentle reminder. Thanks.

File: Makefile [hunk 4]
```diff
@@ -58,7 +58,7 @@ $(TARGET_BIN):
 
 binaries: $(TARGET_BIN)
 
-install:
+install: binaries
 	$(INSTALL) -D -m 755 $(CURDIR)/bin/$(TARGET_BIN) $(BINDIR)/$(TARGET_BIN)
```
Comment (AkihiroSuda) [inline Makefile:62]: This will break `make && sudo make install && make clean`, as `make clean` will be unable to remove the binary that is built during `sudo make install`

Comment (sshedi) [reply Makefile:62]: Sorry, I didn't get you here. It's working fine for me.
```
$ make && make install && make clean
GO111MODULE=on CGO_ENABLED=0 go build -ldflags "-s -w -X github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc/version.Version=v1.0.4-23-gd007ead -X github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc/version.Revision=d007ead4d04d51423aca5d76c7b30cf25f890a1d" -o bin/containerd-fuse-overlayfs-grpc github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc
GO111MODULE=on CGO_ENABLED=0 go build -ldflags "-s -w -X github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc/version.Version=v1.0.4-23-gd007ead -X github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc/version.Revision=d007ead4d04d51423aca5d76c7b30cf25f890a1d" -o /root/fuse-overlayfs-snapshotter/bin/containerd-fuse-overlayfs-grpc github.com/containerd/fuse-overlayfs-snapshotter/cmd/containerd-fuse-overlayfs-grpc
install -D -m 755 /root/fuse-overlayfs-snapshotter/bin/containerd-fuse-overlayfs-grpc /usr/local/bin/containerd-fuse-overlayfs-grpc
rm -f -r /root/fuse-overlayfs-snapshotter/bin /root/fuse-overlayfs-snapshotter/_output
```

File: Makefile [hunk 5]
```diff
@@ -12,17 +12,26 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
+# Deliverables path
```
[the remark this answers is not in the document]
Comment (sshedi) [reply Makefile:15]: The final binaries that get installed.

File: Makefile (hunk 4 above)
Comment (AkihiroSuda) [reply Makefile:62]: That is working because you are running everything as root

Comment (sshedi) [reply Makefile:62]: I checked in a fedora host.
```
➜ $ ls -ld /usr/
drwxr-xr-x. 1 root root 100 Aug 30 10:48 /usr/
[Tue 08.Nov.2022 17:22:31] sshedi@ ~ ➤
➜ $ ls -ld /usr/local
drwxr-xr-x. 1 root root 90 Aug  9 19:27 /usr/local
[Tue 08.Nov.2022 17:22:33] sshedi@ ~ ➤
➜ $ ls -ld /usr/local/bin
drwxr-xr-x. 1 root root 228 Oct 25 13:19 /usr/local/bin
```

So it's evident that regular user can't write to `/usr/local/bin`, using `sudo` is a must.

But `make clean` works fine as regular user, no problems.
```
test@ph4dev [ ~/fuse-overlayfs-snapshotter ]$ ls
bin       cmd         fuseoverlayfs.go       go.mod  LICENSE      Makefile  README.md
check.go  Dockerfile  fuseoverlayfs_test.go  go.sum  MAINTAINERS  plugin

test@ph4dev [ ~/fuse-overlayfs-snapshotter ]$ ls bin
containerd-fuse-overlayfs-grpc

test@ph4dev [ ~/fuse-overlayfs-snapshotter ]$ make clean
rm -f -r /home/test/fuse-overlayfs-snapshotter/bin /home/test/fuse-overlayfs-snapshotter/_output
```

So, I didn't get your expectation here, can you please explain more?

Comment (AkihiroSuda) [reply Makefile:62]: `make && sudo make install && make clean' seems to work, but `sudo make install && make clean` doesn't work for me

```console
$ make clean
rm -f -r /home/suda/gopath/src/github.com/containerd/fuse-overlayfs-snapshotter/bin /home/suda/gopath/src/github.com/containerd/fuse-overlayfs-snapshotter/_output
rm: cannot remove '/home/suda/gopath/src/github.com/containerd/fuse-overlayfs-snapshotter/bin/containerd-fuse-overlayfs-grpc': Permission denied
make: *** [Makefile:68: clean] Error 1
```

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Please drop commit `[Makefile: install should depend on binaries](https://github.com/containerd/fuse-overlayfs-snapshotter/pull/43/commits/d007ead4d04d51423aca5d76c7b30cf25f890a1d)`, then LGTM

File: Makefile (hunk 4 above)
Comment (sshedi) [reply Makefile:62]: Got the problem.
`make && sudo make installl && make clean` this will create binaries as regular user and install as sudo and make clean is no problem.

But with `sudo make install && make clean` (no make at first as regular user), binaries get created as sudo hence regular user can't delete it.

**Comment by sshedi** [comment]
> Please drop commit [`Makefile: install should depend on binaries`](https://github.com/containerd/fuse-overlayfs-snapshotter/pull/43/commits/d007ead4d04d51423aca5d76c7b30cf25f890a1d), then LGTM

Done, thanks for your time and patience.

**Review by AkihiroSuda — APPROVED** [review]
Thanks
